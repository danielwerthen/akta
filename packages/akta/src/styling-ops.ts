type StyleSheetNode = [string, HTMLStyleElement[]];
export function stylesheetMap2() {
  const sheets: StyleSheetNode[] = [['', []]];
  function get(attrName: string, media: string = ''): HTMLStyleElement {
    const specificity = attrName.split('-').length - 1;
    let i = 0;
    let node: StyleSheetNode | null = null;
    for (i = 0; i < sheets.length; i++) {
      if (sheets[i][0] === media) {
        node = sheets[i];
        break;
      }
      if (sheets[i + 1] && sheets[i + 1][0] > media) {
        sheets.splice(i + 1, 0, (node = [media, []]));
        break;
      }
    }
    if (!node) {
      node = [media, []];
      sheets.push(node);
    }
    if (!node[1][specificity]) {
      const sheet = document.createElement('style');
      sheet.media = media;
      sheet.dataset['specificity'] = specificity.toString();
      node[1][specificity] = sheet;
      let prevSheet: HTMLStyleElement | null = null;
      for (let j = i; j >= 0; j--) {
        if (j === i) {
          if (specificity > 0) {
            for (let k = specificity - 1; k >= 0; k--) {
              prevSheet = sheets[j][1][k];
              if (prevSheet) {
                break;
              }
            }
          }
        } else {
          if (specificity > 0) {
            for (let k = sheets[j][1].length; k >= 0; k--) {
              prevSheet = sheets[j][1][k];
              if (prevSheet) {
                break;
              }
            }
          }
        }
        if (prevSheet) {
          break;
        }
      }
      if (!prevSheet) {
        let nextSheet: HTMLStyleElement | null = null;
        for (let j = i; j < sheets.length; j++) {
          for (let k = specificity + 1; k < sheets[j][1].length; k--) {
            nextSheet = sheets[j][1][k];
            if (nextSheet) {
              break;
            }
          }
          if (nextSheet) {
            break;
          }
        }
        if (nextSheet) {
          nextSheet.before(sheet);
        } else {
          document.head.appendChild(sheet);
        }
      } else {
        prevSheet.after(sheet);
      }
    }
    return node[1][specificity];
  }
  return get;
}

function isStyleElement(obj: ChildNode): obj is HTMLStyleElement {
  return typeof (obj as HTMLStyleElement).media === 'string';
}

export function stylesheetMap() {
  let startNode: HTMLStyleElement | undefined;
  function get(attrName: string, media: string = ''): HTMLStyleElement {
    const specificity = (attrName.split('-').length - 1).toString();
    if (!startNode) {
      const sheet = (startNode = document.createElement('style'));
      sheet.media = media;
      sheet.dataset['specificity'] = specificity;
      document.head.append(sheet);
      return sheet;
    }
    let node: HTMLStyleElement | null = startNode;
    while (node) {
      const a = node.media || 'ZZZZ';
      const b = media || 'ZZZZ';
      const spec = node.dataset['specificity'] ?? '0';
      if (node.media === media && spec === specificity) {
        return node;
      }
      if (a < b || (a === b && spec > specificity)) {
        const newNode = document.createElement('style');
        newNode.media = media;
        newNode.dataset['specificity'] = specificity;
        node.before(newNode);
        startNode = newNode;
        return newNode;
      }
      const nextNode = node?.nextSibling as HTMLStyleElement;
      if (!nextNode || !isStyleElement(nextNode)) {
        break;
      }
      node = nextNode;
    }
    const newNode = document.createElement('style');
    newNode.media = media;
    newNode.dataset['specificity'] = specificity;
    node.after(newNode);
    return newNode;
  }
  return get;
}
