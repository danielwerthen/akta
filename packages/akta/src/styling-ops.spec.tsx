import { parseProp, stylesheetMap } from './styling-ops';

describe('styling-ops', () => {
  it('should attach style tags in the right order', () => {
    const get = stylesheetMap();
    get('2');
    get('1');
    get('3');
    get('0');

    get('2', '(max-width: 350px)');
    get('1', '(max-width: 350px)');
    get('4', '(max-width: 350px)');
    get('3', '(max-width: 250px)');
    get('0', '(max-width: 750px)');
    get('4');
    get('4');
    get('4');
    get('4');
    /*
     */
    expect(document.head).toMatchSnapshot();
  });
  it('should properly parse props', () => {
    expect(
      [
        'color',
        'backgroundColor',
        'borderTopWidth',
        'color_max100',
        'color_max100_min40',
        'color_max100_hover',
        'color_max100_hover_focus',
        'color_large',
        'color_nthChild(2n)',
      ].map(parseProp)
    ).toMatchSnapshot();
  });
});
