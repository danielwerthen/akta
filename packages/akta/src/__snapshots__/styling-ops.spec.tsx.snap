// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styling-ops should attach style tags in the right order 1`] = `
<head>
  <style
    data-specificity="0"
    media=""
  />
  <style
    data-specificity="1"
    media=""
  />
  <style
    data-specificity="2"
    media=""
  />
  <style
    data-specificity="3"
    media=""
  />
  <style
    data-specificity="4"
    media=""
  />
  <style
    data-specificity="0"
    media="(max-width: 750px)"
  />
  <style
    data-specificity="1"
    media="(max-width: 350px)"
  />
  <style
    data-specificity="2"
    media="(max-width: 350px)"
  />
  <style
    data-specificity="4"
    media="(max-width: 350px)"
  />
  <style
    data-specificity="3"
    media="(max-width: 250px)"
  />
</head>
`;

exports[`styling-ops should properly parse props 1`] = `
Array [
  Object {
    "attributeName": "color",
    "specificity": 0,
  },
  Object {
    "attributeName": "background-color",
    "specificity": 1,
  },
  Object {
    "attributeName": "border-top-width",
    "specificity": 2,
  },
  Object {
    "attributeName": "color",
    "media": "(max-width: 100px)",
    "specificity": 0,
  },
  Object {
    "attributeName": "color",
    "media": "(max-width: 100px) and (min-width: 40px)",
    "specificity": 0,
  },
  Object {
    "attributeName": "color",
    "media": "(max-width: 100px)",
    "pseudo": ":hover",
    "specificity": 0,
  },
  Object {
    "attributeName": "color",
    "media": "(max-width: 100px)",
    "pseudo": ":hover:focus",
    "specificity": 0,
  },
  Object {
    "attributeName": "color",
    "media": "(min-width: 1201px)",
    "specificity": 0,
  },
  Object {
    "attributeName": "color",
    "pseudo": ":nth-child(2n)",
    "specificity": 0,
  },
]
`;
